### ISSUE TITLE
[Discussion] Roadmap for Pyodide 1.0

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#2304]
Based on the way things are headed, I think we should be able to make a v1.0 release some time in the next year or two. But we have many different components:

1. core Pyodide ffi
2. build system / meta.yaml format
3. micropip
4. matplotlib
5. docker images

Each of these will have different stability guarantees. I think it would be okay for instance if version 1.0 is primarily about the core pyodide ffi and micropip, and maybe we make weaker guarantees about the build system and matplotlib. Ideally the docs could be fairly explicit about which components are less stable. 

We also have:
1. Python version updates, and 
2. Emscripten version updates.
What do these updates mean for our versioning? Even if the `meta.yaml` format / build system is unchanged, updating emscripten may cause differences in behavior. Emscripten has no well-defined ABI, so updating Emscripten versions is a "recompile the world" situation (at least in prinicpal).

It would be good to try to iron out any design issues that are likely to require compatibility breaks sooner rather than later. For instance, I think it would be good to go through the Python apis and look for arguments that could be make positional only or keyword only, similar to #2300.

**Comment by ryanking13** [comment]
Thanks for opening the discussion on this!

> I think it would be okay for instance if version 1.0 is primarily about the core pyodide ffi and micropip, and maybe we make weaker guarantees about the build system and matplotlib. Ideally the docs could be fairly explicit about which components are less stable.

Yeah, I also think we should separate components and have different stability guarantees. For this, unvendoring components from the mono repo (#2014) must proceed. We will also need a proper CLI (#1977) for handling those together.

**Comment by hoodmane** [comment]
I don't think we necessarily need to unvendor anything or even to have separate version numbers. We can just say: This part is very stable. This part is sort of stable. This part is unstable. Many projects have API surfaces with a mixture of stability levels, all with a common version number. If we mark an API as "experimental" then a compatibility break can be a minor version step. This was more the direction I was thinking of going.

I think it would be worth considering a rolling stabilization model where newly added APIs can be marked as unstable and are only stabilized after a few versions. Basically the deprecation model in reverse.

**Comment by ryanking13** [comment]
I'll cancel my word *must*. Yes, we can say some part is stable and other parts are not, regardless of monorepo. Still, I think it's optimal to start unvendoring components someday when the system becomes more stable (and when some of us have more time to do that huge work :/ )
